### PR TITLE
ROC-4726: Settlement agreement events creation

### PR DESCRIPTION
**Pull request by karthikkannankainos**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/ROC-4726

### Change description ###
Settlement agreement and Offer event creation.

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/cmc/claimstore/services/SettlementAgreementService.java [hunk 1]
```diff
@@ -65,18 +74,32 @@ public Claim countersign(Claim claim, String authorisation) {
...
         AppInsightsEvent appInsightsEvent = settlement.isSettlementThroughAdmissions()
             ? SETTLEMENT_AGREEMENT_REACHED_BY_ADMISSION : SETTLEMENT_AGREEMENT_REACHED;
         appInsights.trackEvent(appInsightsEvent, REFERENCE_NUMBER, updated.getReferenceNumber());
 
         return updated;
+
+    }
+
+    protected void signSettlementAgreement(String externalId, Settlement settlement, String authorisation) {
+        final Claim claim = claimService.getClaimByExternalId(externalId, authorisation);
```
Comment (markry-kainos) [inline src/main/java/uk/gov/hmcts/cmc/claimstore/services/SettlementAgreementService.java]: minor: do we need the `final` and `this.` here?

File: ccd-adapter/src/main/java/uk/gov/hmcts/cmc/ccd/mapper/defendant/DefendantMapper.java [hunk 2]
```diff
@@ -67,11 +78,14 @@ public TheirDetails from(Claim.ClaimBuilder builder, CCDDefendant defendant) {
             .defendantId(defendant.getDefendantId());
 
         countyCourtJudgmentMapper.from(defendant.getCountyCourtJudgementRequest(), builder);
+        builder.settlement(settlementMapper.fromCCDDefendant(defendant));
 
         Optional.ofNullable(defendant.getResponseMoreTimeNeededOption()).ifPresent(
             moreTimeNeeded -> builder.moreTimeRequested(moreTimeNeeded.toBoolean())
         );
 
+        builder.settlementReachedAt(defendant.getSettlementReachedAt());
```
Comment (sabahirfan) [inline ccd-adapter/src/main/java/uk/gov/hmcts/cmc/ccd/mapper/defendant/DefendantMapper.java]: minor: Would be nice to have this next to line 81: `builder.settlement(settlementMapper.fromCCDDefendant(defendant));`

File: src/main/java/uk/gov/hmcts/cmc/claimstore/services/SettlementAgreementService.java [hunk 3]
```diff
@@ -31,31 +36,35 @@
...
+        this.ccdEventProducer = ccdEventProducer;
     }
 
     public Claim reject(Claim claim, String authorisation) {
 
         Settlement settlement = assertSettlementCanBeResponded(claim);
         settlement.reject(MadeBy.DEFENDANT);
 
-        String userAction = format("SETTLEMENT_AGREEMENT_REJECTED_BY_%s", MadeBy.DEFENDANT.name());
+        String userAction = AGREEMENT_REJECTED_BY_DEFENDANT.getValue();
```
Comment (sabahirfan) [inline src/main/java/uk/gov/hmcts/cmc/claimstore/services/SettlementAgreementService.java]: Please do not to use CCD event names here in this context. Please revert these changes back.

Comment (karthikkannankainos) [reply src/main/java/uk/gov/hmcts/cmc/claimstore/services/SettlementAgreementService.java]: Not following your comment. I will call you

Comment (dharmendrak) [reply src/main/java/uk/gov/hmcts/cmc/claimstore/services/SettlementAgreementService.java]: sabah, these event names are just for ccd only, on db call it is ignored

File: src/main/java/uk/gov/hmcts/cmc/claimstore/services/SettlementAgreementService.java (hunk 1 above)
Comment (dharmendrak) [reply src/main/java/uk/gov/hmcts/cmc/claimstore/services/SettlementAgreementService.java]: please remove all final from method, we cleaned final usage from every where.

File: ccd-adapter/src/test/java/uk/gov/hmcts/cmc/ccd/mapper/defendant/DefendantMapperTest.java [hunk 4]
```diff
@@ -205,4 +210,39 @@ public void mapCountyCourtJudgmentToCCDDefendant() {
...
+        //When
+        CCDDefendant ccdDefendant = mapper.to(theirDetails, claimWithSettlement);
+
+        //Then
+        assertNotNull(ccdDefendant.getSettlementPartyStatements());
+        assertNotNull(ccdDefendant.getSettlementReachedAt());
+        assertThat(ccdDefendant.getSettlementPartyStatements().size(), is(2));
+        assertEquals(settlementReachedAt, ccdDefendant.getSettlementReachedAt());
+    }
+
```
Comment (sabahirfan) [inline ccd-adapter/src/test/java/uk/gov/hmcts/cmc/ccd/mapper/defendant/DefendantMapperTest.java]: Please add unit test to map back from CCD to CMC for the same.

File: ccd-adapter/src/test/java/uk/gov/hmcts/cmc/ccd/util/SampleCCDDefendant.java [hunk 5]
```diff
@@ -145,4 +149,18 @@ private SampleCCDDefendant() {
                     ).build()
                 ));
     }
+
+    public static CCDDefendant.CCDDefendantBuilder withPartyStatements() {
+        List<CCDCollectionElement<CCDPartyStatement>> partyStatements =
+            Arrays.asList(
```
[the remark this answers is not in the document]
Comment (sabahirfan) [reply ccd-adapter/src/test/java/uk/gov/hmcts/cmc/ccd/util/SampleCCDDefendant.java]: Please clean this

File: src/main/java/uk/gov/hmcts/cmc/claimstore/services/OffersService.java [hunk 6]
```diff
@@ -55,7 +58,9 @@ public Claim makeOffer(Claim claim, Offer offer, MadeBy party, String authorisat
         Settlement settlement = claim.getSettlement().orElse(new Settlement());
         settlement.makeOffer(offer, party);
 
-        String userAction = userAction("OFFER_MADE_BY", party.name());
+        String userAction
+            = party == MadeBy.CLAIMANT ? OFFER_MADE_BY_CLAIMANT.getValue() : OFFER_MADE_BY_DEFENDANT.getValue();
```
Comment (sabahirfan) [inline src/main/java/uk/gov/hmcts/cmc/claimstore/services/OffersService.java]: Can we please use enum itself rather than passing `getValue()`?

File: src/main/java/uk/gov/hmcts/cmc/claimstore/services/SettlementAgreementService.java (hunk 3 above)
Comment (sabahirfan) [reply src/main/java/uk/gov/hmcts/cmc/claimstore/services/SettlementAgreementService.java]: Please ignore above, but try to use enum rather than `getValue()` where ever possible.

Comment (karthikkannankainos) [reply src/main/java/uk/gov/hmcts/cmc/claimstore/services/SettlementAgreementService.java]: @sabahirfan the value matches the events in CCD, currently testing it.

File: src/main/java/uk/gov/hmcts/cmc/claimstore/services/SettlementAgreementService.java (hunk 1 above)
Comment (karthikkannankainos) [reply src/main/java/uk/gov/hmcts/cmc/claimstore/services/SettlementAgreementService.java]: @dharmendrak , @markry-kainos and @sabahirfan this was refactor from offerservice into this service, just lift and shift. (not my code) but Will sort it out.

File: ccd-adapter/src/main/java/uk/gov/hmcts/cmc/ccd/mapper/defendant/DefendantMapper.java [hunk 7]
```diff
@@ -50,6 +54,13 @@ public CCDDefendant to(TheirDetails theirDetails, Claim claim) {
         builder.directionsQuestionnaireDeadline(claim.getDirectionsQuestionnaireDeadline());
         builder.countyCourtJudgementRequest(countyCourtJudgmentMapper.to(claim));
 
+        claim.getSettlement().ifPresent(settlement ->
```
[the remark this answers is not in the document]
Comment (karthikkannankainos) [reply ccd-adapter/src/main/java/uk/gov/hmcts/cmc/ccd/mapper/defendant/DefendantMapper.java]: why do we need claim to be passed here? this is not claimmapper.

File: src/main/java/uk/gov/hmcts/cmc/claimstore/services/SettlementAgreementService.java (hunk 3 above)
Comment (sabahirfan) [reply src/main/java/uk/gov/hmcts/cmc/claimstore/services/SettlementAgreementService.java]: We are passing string value here and then in `CCDCaseRepository` converting it back to enum by calling `CaseEvent.valueOf(userAction)`. Would be clean to pass enum and avoid unnecessary conversion. 

Here we are converting it back to enum.
https://github.com/hmcts/cmc-claim-store/blob/585e6a1491c36fef5b66e1a0d90bdab598cb3a3b/src/main/java/uk/gov/hmcts/cmc/claimstore/repositories/CCDCaseRepository.java#L135

Comment (karthikkannankainos) [reply src/main/java/uk/gov/hmcts/cmc/claimstore/services/SettlementAgreementService.java]: @sabahirfan I saw this already. I am already refactoring it to send it as Enum.

File: ccd-adapter/src/main/java/uk/gov/hmcts/cmc/ccd/mapper/defendant/DefendantMapper.java [hunk 8]
```diff
@@ -67,11 +78,14 @@ public TheirDetails from(Claim.ClaimBuilder builder, CCDDefendant defendant) {
             .defendantId(defendant.getDefendantId());
 
         countyCourtJudgmentMapper.from(defendant.getCountyCourtJudgementRequest(), builder);
+        builder.settlement(settlementMapper.fromCCDDefendant(defendant));
```
[the remark this answers is not in the document]
Comment (karthikkannankainos) [reply ccd-adapter/src/main/java/uk/gov/hmcts/cmc/ccd/mapper/defendant/DefendantMapper.java]: The settlementmapper maps/creates settlement from information. While this DefendantMapper aggregates all information what it needs from Defendant.

File: ccd-adapter/src/main/java/uk/gov/hmcts/cmc/ccd/mapper/defendant/DefendantMapper.java (hunk 2 above)
Comment (karthikkannankainos) [reply ccd-adapter/src/main/java/uk/gov/hmcts/cmc/ccd/mapper/defendant/DefendantMapper.java]: done

File: ccd-adapter/src/main/java/uk/gov/hmcts/cmc/ccd/mapper/offers/SettlementMapper.java [hunk 9]
```diff
@@ -24,23 +22,24 @@ public SettlementMapper(PartyStatementMapper partyStatementMapper) {
         this.partyStatementMapper = partyStatementMapper;
     }
 
-    @Override
-    public CCDSettlement to(Settlement settlement) {
-        CCDSettlement.CCDSettlementBuilder builder = CCDSettlement.builder();
-
-        List<CCDCollectionElement<CCDPartyStatement>> partyStatements = settlement.getPartyStatements().stream()
+    public List<CCDCollectionElement<CCDPartyStatement>> toCCDPartyStatements(Settlement settlement) {
```
[the remark this answers is not in the document]
Comment (karthikkannankainos) [reply ccd-adapter/src/main/java/uk/gov/hmcts/cmc/ccd/mapper/offers/SettlementMapper.java]: we dont have one consistent way or agreed design principle in mappers. We moved away from mapper interface too. So whatever is best, have been chosen here.

File: ccd-adapter/src/test/java/uk/gov/hmcts/cmc/ccd/mapper/defendant/DefendantMapperTest.java (hunk 4 above)
Comment (karthikkannankainos) [reply ccd-adapter/src/test/java/uk/gov/hmcts/cmc/ccd/mapper/defendant/DefendantMapperTest.java]: done
